File: modesti-server/src/main/frontend/app/schema/category/category.ts
import {Conditional} from '../conditional';
import {Field} from '../field/field';

export class Category {
  public id: string;
  public name: string;
  public editable: Conditional;
  public fields: Field[];

  public deserialize(category: Category): Category {
    this.id = category.id;
    this.name = category.name;
    this.editable = category.editable;

    if (category.fields) {
      this.fields = [];
      category.fields.forEach((field: Field) => {
        this.fields.push(new Field().deserialize(field));
      });
    }

    return this;
  }
}

File: modesti-api/src/main/java/cern/modesti/schema/category/Category.java
package cern.modesti.schema.category;

import cern.modesti.schema.field.Field;

import java.util.Map;
import java.io.Serializable;
import java.util.List;

/**
 * @author Justin Lewis Salmon
 */
public interface Category extends Serializable {

  String getId();

  void setId(String id);

  String getName();

  void setName(String name);

  List<Field> getFields();

  List<Field> getFields(List<String> fieldIds);

  void setFields(List<Field> fields);

  List<String> getExcludes();

  void setExcludes(List<String> excludes);

  List<Constraint> getConstraints();

  void setConstraints(List<Constraint> constraints);

  Map<String, Object> getEditable();

  void setEditable(Map<String, Object> editables);

  List<String> getFieldNames(List<String> fieldIds);
}
